File: src/ui/content/contentState.ts
import {
  Id, StringSlice, Sentence, NodeIndicatorInPlot, PlotIndex
} from '../../content/types';
import {
  adoptNodesInTree, disownNodesInTree,
  InsertedNode, insertNodeIntoTree, transformNodeInTree, transformNodesInTree
} from '../../content/unpositioned/manipulation';
import { deleteNodesInPlot, transformNodesInPlot } from '../../content/unpositioned/plotManipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../../util/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { changeAt, insertAt, omitKey, removeAt } from '../../util/objTransforms';
import {
  isTerminal, PlotCoordsOffset, UnpositionedPlot, UnpositionedTree
} from '../../content/unpositioned/types';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'addPlot' }
  | { type: 'deletePlot', plotIndex: PlotIndex }
  | { type: 'insertNode', plotIndex: PlotIndex, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'adoptNodes', plotIndex: PlotIndex, treeId: Id, adoptingNodeId: Id, adoptedNodeIds: Id[] }
  | { type: 'disownNodes', plotIndex: PlotIndex, treeId: Id, disowningNodeId: Id, disownedNodeIds: Id[] }
  | { type: 'moveNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], dx: number, dy: number }
  | { type: 'resetNodePositions', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'setNodeLabel', plotIndex: PlotIndex, nodeIndicator: NodeIndicatorInPlot, newLabel: string }
  | { type: 'setTriangle', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], triangle: boolean }
  | { type: 'setSentence', plotIndex: PlotIndex, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setPlot', plotIndex: PlotIndex, old: UnpositionedPlot, new: UnpositionedPlot }
  | { type: 'addPlot', newPlotIndex: PlotIndex, newPlot: UnpositionedPlot }
  | { type: 'removePlot', plotIndex: PlotIndex, removedPlot: UnpositionedPlot }
  | { type: 'setTree', plotIndex: PlotIndex, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'addPlot': {
      return {
        type: 'addPlot',
        newPlotIndex: state.plots.length,
        newPlot: { trees: {} },
      };
    }
    case 'deletePlot': {
      return {
        type: 'removePlot',
        plotIndex: action.plotIndex,
        removedPlot: state.plots[action.plotIndex],
      }
    }
    case 'insertNode': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'deleteNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: deleteNodesInPlot(action.nodeIndicators)(state.plots[action.plotIndex]),
      };
    }
    case 'adoptNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: adoptNodesInTree(action.adoptingNodeId, action.adoptedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'disownNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: disownNodesInTree(action.disowningNodeId, action.disownedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'moveNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({
          ...node,
          offset: { dTreeX: node.offset.dTreeX + action.dx, dTreeY: node.offset.dTreeY + action.dy }
        }))(action.nodeIndicators)(state.plots[action.plotIndex]),
      }
    }
    case 'resetNodePositions': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({ ...node, offset: { dTreeX: 0, dTreeY: 0 } }))(action.nodeIndicators)(
          state.plots[action.plotIndex]),
      };
    }
    case 'setNodeLabel': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.nodeIndicator.treeId,
        old: state.plots[action.plotIndex].trees[action.nodeIndicator.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.nodeIndicator.nodeId)(
          state.plots[action.plotIndex].trees[action.nodeIndicator.treeId]),
      };
    }
    case 'setTriangle': {
      const treeId = action.nodeIndicators[0].treeId;  // TODO: Use all trees
      const nodeIds = action.nodeIndicators
        .filter(nodeIndicator => nodeIndicator.treeId === treeId).map(nodeIndicator => nodeIndicator.nodeId);
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId,
        old: state.plots[action.plotIndex].trees[treeId],
        new: transformNodesInTree(node => isTerminal(node) ? { ...node, triangle: action.triangle } : node)(nodeIds)(
          state.plots[action.plotIndex].trees[treeId]),
      };
    }
    case 'setSentence': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'addTree': {
      return {
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    }
    case 'removeTree': {
      return {
        ...action,
        removedTree: state.plots[action.plotIndex].trees[action.treeId],
      };
    }
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, action.new),
      };
    }
    case 'addPlot': {
      return {
        ...state,
        plots: insertAt(state.plots, action.newPlotIndex, action.newPlot),
      };
    }
    case 'removePlot': {
      return {
        ...state,
        plots: removeAt(state.plots, action.plotIndex),
      }
    }
    case 'setTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.treeId]: action.new,
          },
        }),
      };
    }
    case 'addTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.newTreeId]: action.newTree,
          },
        }),
      };
    }
    case 'removeTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: omitKey(state.plots[action.plotIndex].trees, action.treeId)
        }),
      };
    }
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addPlot': {
      return {
        ...action,
        type: 'removePlot',
        plotIndex: action.newPlotIndex,
        removedPlot: action.newPlot || { trees: {} },
      };
    }
    case 'removePlot': {
      return {
        ...action,
        type: 'addPlot',
        newPlotIndex: action.plotIndex,
        newPlot: action.removedPlot,
      }
    }
    case 'setTree': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addTree': {
      return {
        ...action,
        type: 'removeTree',
        plotIndex: action.plotIndex,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    }
    case 'removeTree': {
      return {
        ...action,
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    }
    default:
      return action;
  }
};

export type ContentState = {
  plots: UnpositionedPlot[];
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: [
    {
      trees: {},
    },
  ],
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);

File: npbloom-core/src/commonMain/kotlin/space/yuvalinguist/npbloom/ui/content/ContentState.kt
@file:OptIn(ExperimentalJsExport::class)

package space.yuvalinguist.npbloom.ui.content

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import space.yuvalinguist.npbloom.changeAt
import space.yuvalinguist.npbloom.content.*
import space.yuvalinguist.npbloom.content.unpositioned.*
import space.yuvalinguist.npbloom.history.UndoRedoHistory
import space.yuvalinguist.npbloom.history.UndoableActionBase
import space.yuvalinguist.npbloom.insertAt
import space.yuvalinguist.npbloom.removeAt
import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport
import kotlin.js.JsName

private typealias PlotIndex = Int

@JsExport
@Serializable
data class ContentState(
    @SerialName("p") @JsName("plotsAsKtList") val plots: List<UnpositionedPlot>
) {
    @JsName("plots") val plotsAsArray get() = plots.toTypedArray()
}

private val initialState = ContentState(listOf(UnpositionedPlot()))

internal sealed interface ContentOrHistoryAction

data object Undo : ContentOrHistoryAction
data object Redo : ContentOrHistoryAction

private sealed interface ContentAction : UndoableActionBase, ContentOrHistoryAction

data object AddPlot : ContentAction
internal data class DeletePlot(val plotIndex: PlotIndex) : ContentAction
internal data class ResetPlot(val plotIndex: PlotIndex) : ContentAction
internal data class InsertNode(val plotIndex: PlotIndex, val treeId: Id, val newNode: InsertedNode) : ContentAction
internal data class DeleteNodes(val plotIndex: PlotIndex, val nodeIndicators: Set<NodeIndicatorInPlot>) : ContentAction
internal data class AdoptNodes(
    val plotIndex: PlotIndex,
    val treeId: Id,
    val adoptingNodeId: Id,
    val adoptedNodeIds: Set<Id>
) : ContentAction

internal data class DisownNodes(
    val plotIndex: PlotIndex,
    val treeId: Id,
    val disowningNodeId: Id,
    val disownedNodeIds: Set<Id>
) : ContentAction

internal data class MoveNodes(
    val plotIndex: PlotIndex,
    val nodeIndicators: Set<NodeIndicatorInPlot>,
    val offsetD: TreeCoordsOffset
) : ContentAction

internal data class ResetNodePositions(val plotIndex: PlotIndex, val nodeIndicators: Set<NodeIndicatorInPlot>) :
    ContentAction

internal data class SetNodeLabel(
    val plotIndex: PlotIndex,
    val nodeIndicator: NodeIndicatorInPlot,
    val newLabel: String
) : ContentAction

internal data class SetTriangle(
    val plotIndex: PlotIndex,
    val nodeIndicators: Set<NodeIndicatorInPlot>,
    val triangle: Boolean
) : ContentAction

internal data class SetSentence(
    val plotIndex: PlotIndex,
    val treeId: Id,
    val newSentence: Sentence,
    val oldSelectedSlice: StringSlice
) : ContentAction

internal data class AddTree(val plotIndex: PlotIndex, val newTreeId: Id, val offset: PlotCoordsOffset) : ContentAction
internal data class DeleteTree(val plotIndex: PlotIndex, val treeId: Id) : ContentAction

@JsExport
sealed interface ContentChange : UndoableActionBase

internal data class PlotAdded(val newPlotIndex: PlotIndex, val newPlot: UnpositionedPlot) : ContentChange
internal data class PlotChanged(val plotIndex: PlotIndex, val old: UnpositionedPlot, val new: UnpositionedPlot) :
    ContentChange

internal data class PlotDeleted(val plotIndex: PlotIndex, val removedPlot: UnpositionedPlot) : ContentChange
internal data class TreeAdded(val plotIndex: PlotIndex, val newTree: UnpositionedTree) : ContentChange
internal data class TreeChanged(val plotIndex: PlotIndex, val old: UnpositionedTree, val new: UnpositionedTree) :
    ContentChange

internal data class TreeDeleted(val plotIndex: PlotIndex, val removedTree: UnpositionedTree) : ContentChange

private fun makeUndoable(state: ContentState, action: ContentAction): ContentChange = when (action) {
    AddPlot -> PlotAdded(state.plots.size, UnpositionedPlot())
    is DeletePlot -> PlotDeleted(action.plotIndex, state.plots[action.plotIndex])
    is ResetPlot -> PlotChanged(action.plotIndex, state.plots[action.plotIndex], UnpositionedPlot())
    is InsertNode -> TreeChanged(
        action.plotIndex,
        state.plots[action.plotIndex].tree(action.treeId),
        state.plots[action.plotIndex].tree(action.treeId).insertNode(action.newNode)
    )

    is DeleteNodes -> PlotChanged(
        action.plotIndex,
        state.plots[action.plotIndex],
        state.plots[action.plotIndex].deleteNodes(action.nodeIndicators)
    )

    is AdoptNodes -> TreeChanged(
        action.plotIndex,
        state.plots[action.plotIndex].tree(action.treeId),
        state.plots[action.plotIndex].tree(action.treeId).adoptNodes(action.adoptingNodeId, action.adoptedNodeIds)
    )

    is DisownNodes -> TreeChanged(
        action.plotIndex,
        state.plots[action.plotIndex].tree(action.treeId),
        state.plots[action.plotIndex].tree(action.treeId).disownNodes(action.disowningNodeId, action.disownedNodeIds)
    )

    is MoveNodes -> PlotChanged(
        action.plotIndex,
        state.plots[action.plotIndex],
        state.plots[action.plotIndex].transformNodes(action.nodeIndicators) { it.changeOffset(action.offsetD) }
    )

    is ResetNodePositions -> PlotChanged(
        action.plotIndex,
        state.plots[action.plotIndex],
        state.plots[action.plotIndex].transformNodes(action.nodeIndicators) { it.withOffset(TreeCoordsOffset.ZERO) }
    )

    is SetNodeLabel -> TreeChanged(
        action.plotIndex,
        state.plots[action.plotIndex].tree(action.nodeIndicator.treeId),
        state.plots[action.plotIndex].tree(action.nodeIndicator.treeId)
            .transformNode(action.nodeIndicator.nodeId) { it.withLabel(action.newLabel) }
    )

    is SetTriangle -> PlotChanged(
        action.plotIndex,
        state.plots[action.plotIndex],
        state.plots[action.plotIndex].transformNodes(
            action.nodeIndicators
        ) { if (it is UnpositionedTerminalNode) it.copy(triangle = action.triangle) else it }
    )

    is SetSentence -> TreeChanged(
        action.plotIndex,
        state.plots[action.plotIndex].tree(action.treeId),
        state.plots[action.plotIndex].tree(action.treeId)
            .handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)
    )

    is AddTree -> TreeAdded(action.plotIndex, UnpositionedTree(action.newTreeId, "", EntitySet(), action.offset))
    is DeleteTree -> TreeDeleted(action.plotIndex, state.plots[action.plotIndex].tree(action.treeId))
}

private fun applyUndoableAction(state: ContentState, action: ContentChange): ContentState = when (action) {
    is PlotAdded -> state.copy(plots = state.plots.insertAt(action.newPlotIndex, action.newPlot))
    is PlotChanged -> state.copy(plots = state.plots.changeAt(action.plotIndex, action.new))
    is PlotDeleted -> state.copy(plots = state.plots.removeAt(action.plotIndex))
    is TreeAdded -> state.copy(
        plots = state.plots.changeAt(
            action.plotIndex,
            state.plots[action.plotIndex].setTree(action.newTree)
        )
    )

    is TreeChanged -> state.copy(
        plots = state.plots.changeAt(
            action.plotIndex,
            state.plots[action.plotIndex].setTree(action.new)
        )
    )

    is TreeDeleted -> state.copy(
        plots = state.plots.changeAt(
            action.plotIndex,
            state.plots[action.plotIndex].removeTree(action.removedTree.id)
        )
    )
}

private fun reverseUndoableAction(action: ContentChange): ContentChange = when (action) {
    is PlotAdded -> PlotDeleted(action.newPlotIndex, action.newPlot)
    is PlotChanged -> action.copy(old = action.new, new = action.old)
    is PlotDeleted -> PlotAdded(action.plotIndex, action.removedPlot)
    is TreeAdded -> TreeDeleted(action.plotIndex, action.newTree)
    is TreeChanged -> action.copy(old = action.new, new = action.old)
    is TreeDeleted -> TreeAdded(action.plotIndex, action.removedTree)
}

typealias UndoableContentState = UndoRedoHistory<ContentState, ContentChange>

internal val initialContentState: UndoableContentState =
    UndoRedoHistory(::applyUndoableAction, ::reverseUndoableAction, initialState)

internal fun contentReducer(state: UndoableContentState, action: ContentOrHistoryAction): UndoableContentState =
    when (action) {
        Undo -> state.undo()
        Redo -> state.redo()
        is ContentAction -> state.applyAction(makeUndoable(state.current, action))
    }
